File: src/so/aporia/utils/files/AprParser.java
/*
 * Copyright (c) 2025-2026 Aporia.cc Project
 * Distributed under the Aporia.cc Software License Agreement v1.0
 * See LICENSE and COPYRIGHT files in the project root for full text.
 */

package so.aporia.utils.files;

import java.util.*;
import java.util.regex.*;

/**
 * Parser for .apr config language.
 * <p>
 * Syntax:
 * <pre>
 * module 'Name' (
 *   setting 'key' = T, F
 *   bind = SHIFT
 *   number 'key' = 1.5
 *   text 'key' = 'hello'
 *   select 'key' = 'option1'
 * )
 * </pre>
 */
public final class AprParser {

    private AprParser() {}

    /** Parsed representation of a module config block. */
    public static final class ModuleConfig {
        public final String name;
        public final Map<String, String> settings = new LinkedHashMap<>();
        public int bind = -1;
        public boolean active = false;

        public ModuleConfig(String name) { this.name = name; }
    }

    /** Parsed representation of a full config file. */
    public static final class ConfigFile {
        public final List<ModuleConfig> modules = new ArrayList<>();
    }

    /**
     * Parses .apr config language text into a ConfigFile.
     */
    public static ConfigFile parse(String input) {
        ConfigFile result = new ConfigFile();
        String text = stripComments(input);

        Pattern modulePattern = Pattern.compile(
            "module\\s+'([^']+)'\\s*\\(",
            Pattern.MULTILINE
        );
        Matcher m = modulePattern.matcher(text);

        while (m.find()) {
            String modName = m.group(1);
            int start = m.end();
            int end = findMatchingParen(text, start);
            if (end < 0) continue;

            String body = text.substring(start, end).trim();
            ModuleConfig mc = new ModuleConfig(modName);
            parseBody(body, mc);
            result.modules.add(mc);
        }

        return result;
    }

    /**
     * Serializes a ConfigFile back into .apr language text.
     */
    public static String serialize(ConfigFile cf) {
        StringBuilder sb = new StringBuilder();
        for (ModuleConfig mc : cf.modules) {
            sb.append("module '").append(mc.name).append("' (\n");
            sb.append("  active = ").append(mc.active ? "T" : "F").append("\n");
            if (mc.bind >= 0) {
                sb.append("  bind = ").append(keyToString(mc.bind)).append("\n");
            }
            for (Map.Entry<String, String> e : mc.settings.entrySet()) {
                sb.append("  setting '").append(e.getKey()).append("' = ").append(e.getValue()).append("\n");
            }
            sb.append(")\n\n");
        }
        return sb.toString();
    }

    private static void parseBody(String body, ModuleConfig mc) {
        String[] lines = body.split("\n");
        for (String line : lines) {
            line = line.trim();
            if (line.isEmpty() || line.startsWith("#")) continue;

            // active = T, F
            Pattern activePattern = Pattern.compile("^active\\s*=\\s*(.+)$");
            Matcher am = activePattern.matcher(line);
            if (am.find()) {
                mc.active = am.group(1).trim().equalsIgnoreCase("T") || am.group(1).trim().equalsIgnoreCase("true");
                continue;
            }

            // bind = KEY
            Pattern bindPattern = Pattern.compile("^bind\\s*=\\s*(.+)$");
            Matcher bm = bindPattern.matcher(line);
            if (bm.find()) {
                mc.bind = parseKey(bm.group(1).trim());
                continue;
            }

            // setting 'key' = value
            Pattern settingPattern = Pattern.compile("^setting\\s+'([^']+)'\\s*=\\s*(.+)$");
            Matcher sm = settingPattern.matcher(line);
            if (sm.find()) {
                String key = sm.group(1);
                String val = sm.group(2).trim();
                mc.settings.put(key, val);
            }
        }
    }

    private static int findMatchingParen(String text, int start) {
        int depth = 1;
        for (int i = start; i < text.length(); i++) {
            char c = text.charAt(i);
            if (c == '(') depth++;
            else if (c == ')') {
                depth--;
                if (depth == 0) return i;
            }
        }
        return -1;
    }

    private static String stripComments(String input) {
        return input.replaceAll("#[^\n]*", "");
    }

    /**
     * Converts a GLFW scancode to a string representation.
     */
    public static String keyToString(int scancode) {
        return so.aporia.utils.user.input.KeyCodeMap.getName(scancode);
    }

    /**
     * Parses a key string back to a GLFW scancode.
     */
    public static int parseKey(String keyStr) {
        if (keyStr.equalsIgnoreCase("none") || keyStr.equalsIgnoreCase("null")) return -1;
        return so.aporia.utils.user.input.KeyCodeMap.getScancode(keyStr);
    }
}

File: src/so/aporia/utils/files/AprParser.kt
package so.aporia.utils.files

import com.chaos.annotation.Obfuscate
import so.aporia.utils.user.input.KeyCodeMap
import java.util.LinkedHashMap
import java.util.regex.Pattern

@Obfuscate
object AprParser {

    class ModuleConfig(@JvmField val name: String) {
        @JvmField val settings = linkedMapOf<String, String>()
        @JvmField var bind = -1
        @JvmField var active = false
    }

    class ConfigFile {
        @JvmField val modules = mutableListOf<ModuleConfig>()
    }

    @JvmStatic
    fun parse(input: String): ConfigFile {
        val result = ConfigFile()
        val text = stripComments(input)

        val modulePattern = Pattern.compile("module\\s+'([^']+)'\\s*\\(", Pattern.MULTILINE)
        val m = modulePattern.matcher(text)

        while (m.find()) {
            val modName = m.group(1)
            val start = m.end()
            val end = findMatchingParen(text, start)
            if (end < 0) continue

            val body = text.substring(start, end).trim()
            val mc = ModuleConfig(modName)
            parseBody(body, mc)
            result.modules.add(mc)
        }

        return result
    }

    @JvmStatic
    fun serialize(cf: ConfigFile): String {
        val sb = StringBuilder()
        for (mc in cf.modules) {
            sb.append("module '").append(mc.name).append("' (\n")
            sb.append("  active = ").append(if (mc.active) "T" else "F").append("\n")
            if (mc.bind >= 0) {
                sb.append("  bind = ").append(keyToString(mc.bind)).append("\n")
            }
            for ((key, value) in mc.settings) {
                sb.append("  setting '").append(key).append("' = ").append(value).append("\n")
            }
            sb.append(")\n\n")
        }
        return sb.toString()
    }

    private fun parseBody(body: String, mc: ModuleConfig) {
        for (line in body.split("\n")) {
            val trimmed = line.trim()
            if (trimmed.isEmpty() || trimmed.startsWith("#")) continue

            var m = Pattern.compile("^active\\s*=\\s*(.+)$").matcher(trimmed)
            if (m.find()) {
                mc.active = m.group(1)!!.trim().equals("T", ignoreCase = true) || m.group(1)!!.trim().equals("true", ignoreCase = true)
                continue
            }

            m = Pattern.compile("^bind\\s*=\\s*(.+)$").matcher(trimmed)
            if (m.find()) {
                mc.bind = parseKey(m.group(1)!!.trim())
                continue
            }

            m = Pattern.compile("^setting\\s+'([^']+)'\\s*=\\s*(.+)$").matcher(trimmed)
            if (m.find()) {
                mc.settings[m.group(1)!!] = m.group(2)!!.trim()
            }
        }
    }

    private fun findMatchingParen(text: String, start: Int): Int {
        var depth = 1
        for (i in start until text.length) {
            val c = text[i]
            if (c == '(') depth++
            else if (c == ')') {
                depth--
                if (depth == 0) return i
            }
        }
        return -1
    }

    private fun stripComments(input: String): String = input.replace("#[^\n]*".toRegex(), "")

    @JvmStatic
    fun keyToString(scancode: Int): String = KeyCodeMap.getName(scancode)

    @JvmStatic
    fun parseKey(keyStr: String): Int {
        if (keyStr.equals("none", ignoreCase = true) || keyStr.equals("null", ignoreCase = true)) return -1
        return KeyCodeMap.getScancode(keyStr)
    }
}
